sled-agent/config-reconciler/src/reconciler_task/external_disks.rs

```diff
@@ -0,0 +1,1020 @@
+                "Observed an unexpected zpool uuid";
+                "disk_identity" => ?config.identity,
+                InlineErrorChain::new(&err),
+            );
+            return ExternalDiskState::failed(config, err);
+        }
+
+        // Update any properties that have changed from `disk` based on the
+        // current `raw_disk`. We don't do anything different whether or not any
+        // changes were actually made.
+        let disk = match update_properties_from_raw_disk(disk, raw_disk, log) {
+            MaybeUpdatedDisk::Updated(disk) => disk,
+            MaybeUpdatedDisk::Unchanged(disk) => disk,
+        };
+
+        ExternalDiskState::managed(config, disk)
+    }
+
+    async fn start_managing_disk<T: DiskAdopter>(
+        &self,
+        config: OmicronPhysicalDiskConfig,
+        raw_disk: RawDisk,
+        disk_adopter: &mut T,
+        log: &Logger,
+    ) -> ExternalDiskState {
+        match disk_adopter
+            .adopt_disk(raw_disk, &self.mount_config, config.pool_id, log)
+            .await
+        {
+            Ok(disk) => {
+                info!(
+                    log, "Successfully started management of disk";
+                    "disk_identity" => ?config.identity,
+                );
+                ExternalDiskState::managed(config, disk)
+            }
+            Err(err) => {
+                warn!(
+                    log, "Disk adoption failed";
+                    "disk_identity" => ?config.identity,
+                    InlineErrorChain::new(&err),
+                );
+                ExternalDiskState::failed(config, err)
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+struct ExternalDiskState {
+    config: OmicronPhysicalDiskConfig,
+    state: DiskState,
+}
+
+impl ExternalDiskState {
+    fn managed(config: OmicronPhysicalDiskConfig, disk: Disk) -> Self {
+        Self { config, state: DiskState::Managed(disk) }
+    }
+
+    fn failed(
+        config: OmicronPhysicalDiskConfig,
+        err: DiskManagementError,
+    ) -> Self {
+        Self { config, state: DiskState::FailedToManage(err) }
+    }
+}
+
+impl IdMappable for ExternalDiskState {
+    type Id = PhysicalDiskUuid;
+
+    fn id(&self) -> Self::Id {
+        self.config.id
+    }
+}
+
+#[derive(Debug)]
+enum DiskState {
+    Managed(Disk),
+    FailedToManage(DiskManagementError),
+}
+
+/// Helper to allow unit tests to run without interacting with the real [`Disk`]
+/// implementation. In production, the only implementor of this trait is
+/// [`RealDiskAdopter`].
+trait DiskAdopter {
+    fn adopt_disk(
+        &mut self,
+        raw_disk: RawDisk,
+        mount_config: &MountConfig,
+        pool_id: ZpoolUuid,
+        log: &Logger,
+    ) -> impl Future<Output = Result<Disk, DiskManagementError>> + Send;
+}
+
+struct RealDiskAdopter<'a> {
+    key_requester: &'a StorageKeyRequester,
+}
+
+impl DiskAdopter for RealDiskAdopter<'_> {
+    async fn adopt_disk(
+        &mut self,
```

Comment:
Does this need to be `&mut self`? Or is that just a convenience for tests?

Reply:
It was indeed just a convenience; changed in da04aee4f